File: facturacion-back/facturacion-back/src/main/java/com/cibercom/facturacion_back/service/TiendaService.java
package com.cibercom.facturacion_back.service;

import com.cibercom.facturacion_back.dto.TiendaDto;
import com.cibercom.facturacion_back.model.Tienda;
import com.cibercom.facturacion_back.repository.TiendaRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Servicio para la gestión de tiendas
 */
@Service
@Transactional
public class TiendaService {

    private static final Logger logger = LoggerFactory.getLogger(TiendaService.class);

    @Autowired
    private TiendaRepository tiendaRepository;

    /**
     * Crear una nueva tienda
     */
    public TiendaDto crearTienda(TiendaDto tiendaDto) {
        logger.info("Creando nueva tienda con código: {}", tiendaDto.getCodigoTienda());

        // Verificar si el código ya existe
        if (tiendaRepository.existsByCodigoTienda(tiendaDto.getCodigoTienda())) {
            throw new RuntimeException("Ya existe una tienda con el código: " + tiendaDto.getCodigoTienda());
        }

        Tienda tienda = convertirDtoAEntidad(tiendaDto);
        tienda.setUsuarioCreacion(tiendaDto.getUsuarioCreacion() != null ? 
                                 tiendaDto.getUsuarioCreacion() : "sistema");
        
        Tienda tiendaGuardada = tiendaRepository.save(tienda);
        logger.info("Tienda creada exitosamente con ID: {}", tiendaGuardada.getIdTienda());
        
        return convertirEntidadADto(tiendaGuardada);
    }

    /**
     * Actualizar una tienda existente
     */
    public TiendaDto actualizarTienda(Long idTienda, TiendaDto tiendaDto) {
        logger.info("Actualizando tienda con ID: {}", idTienda);

        Optional<Tienda> tiendaExistente = tiendaRepository.findById(idTienda);
        if (tiendaExistente.isEmpty()) {
            throw new RuntimeException("No se encontró la tienda con ID: " + idTienda);
        }

        // Verificar si el código ya existe en otra tienda
        if (tiendaRepository.existsByCodigoTiendaAndIdTiendaNot(tiendaDto.getCodigoTienda(), idTienda)) {
            throw new RuntimeException("Ya existe otra tienda con el código: " + tiendaDto.getCodigoTienda());
        }

        Tienda tienda = tiendaExistente.get();
        actualizarCamposTienda(tienda, tiendaDto);
        tienda.setUsuarioModificacion(tiendaDto.getUsuarioModificacion() != null ? 
                                     tiendaDto.getUsuarioModificacion() : "sistema");
        tienda.setFechaModificacion(LocalDateTime.now());

        Tienda tiendaActualizada = tiendaRepository.save(tienda);
        logger.info("Tienda actualizada exitosamente con ID: {}", tiendaActualizada.getIdTienda());
        
        return convertirEntidadADto(tiendaActualizada);
    }

    /**
     * Obtener todas las tiendas con filtros opcionales
     */
    @Transactional(readOnly = true)
    public List<TiendaDto> listarTiendas(String estadoTienda, String region, String zona, 
                                        String tipoTienda, String busqueda) {
        logger.info("Listando tiendas con filtros - Estado: {}, Región: {}, Zona: {}, Tipo: {}, Búsqueda: {}", 
                   estadoTienda, region, zona, tipoTienda, busqueda);

        List<Tienda> tiendas = tiendaRepository.findTiendasConFiltros(
                estadoTienda, region, zona, tipoTienda, busqueda);
        
        return tiendas.stream()
                     .map(this::convertirEntidadADto)
                     .collect(Collectors.toList());
    }

    /**
     * Obtener una tienda por ID
     */
    @Transactional(readOnly = true)
    public TiendaDto obtenerTiendaPorId(Long idTienda) {
        logger.info("Obteniendo tienda con ID: {}", idTienda);

        Optional<Tienda> tienda = tiendaRepository.findById(idTienda);
        if (tienda.isEmpty()) {
            throw new RuntimeException("No se encontró la tienda con ID: " + idTienda);
        }

        return convertirEntidadADto(tienda.get());
    }

    /**
     * Obtener una tienda por código
     */
    @Transactional(readOnly = true)
    public TiendaDto obtenerTiendaPorCodigo(String codigoTienda) {
        logger.info("Obteniendo tienda con código: {}", codigoTienda);

        Optional<Tienda> tienda = tiendaRepository.findByCodigoTienda(codigoTienda);
        if (tienda.isEmpty()) {
            throw new RuntimeException("No se encontró la tienda con código: " + codigoTienda);
        }

        return convertirEntidadADto(tienda.get());
    }

    /**
     * Eliminar una tienda (soft delete)
     */
    public void eliminarTienda(Long idTienda, String usuarioModificacion) {
        logger.info("Eliminando tienda con ID: {}", idTienda);

        Optional<Tienda> tiendaExistente = tiendaRepository.findById(idTienda);
        if (tiendaExistente.isEmpty()) {
            throw new RuntimeException("No se encontró la tienda con ID: " + idTienda);
        }

        Tienda tienda = tiendaExistente.get();
        tienda.setEstadoTienda("INACTIVO");
        tienda.setUsuarioModificacion(usuarioModificacion != null ? usuarioModificacion : "sistema");
        tienda.setFechaModificacion(LocalDateTime.now());

        tiendaRepository.save(tienda);
        logger.info("Tienda eliminada (soft delete) exitosamente con ID: {}", idTienda);
    }

    /**
     * Eliminar una tienda permanentemente
     */
    public void eliminarTiendaPermanente(Long idTienda) {
        logger.info("Eliminando permanentemente tienda con ID: {}", idTienda);

        if (!tiendaRepository.existsById(idTienda)) {
            throw new RuntimeException("No se encontró la tienda con ID: " + idTienda);
        }

        tiendaRepository.deleteById(idTienda);
        logger.info("Tienda eliminada permanentemente con ID: {}", idTienda);
    }

    /**
     * Obtener estadísticas de tiendas
     */
    @Transactional(readOnly = true)
    public Map<String, Object> obtenerEstadisticas() {
        logger.info("Obteniendo estadísticas de tiendas");

        Object[] estadisticas = tiendaRepository.getEstadisticasTiendas();
        
        Map<String, Object> resultado = new HashMap<>();
        if (estadisticas != null && estadisticas.length > 0) {
            resultado.put("totalTiendas", estadisticas[0]);
            resultado.put("tiendasActivas", estadisticas[1]);
            resultado.put("tiendasInactivas", estadisticas[2]);
            resultado.put("tiendasSuspendidas", estadisticas[3]);
        }

        // Obtener listas de valores únicos
        resultado.put("regiones", tiendaRepository.findDistinctRegiones());
        resultado.put("zonas", tiendaRepository.findDistinctZonas());
        resultado.put("tiposTienda", tiendaRepository.findDistinctTiposTienda());

        return resultado;
    }

    /**
     * Obtener tiendas activas
     */
    @Transactional(readOnly = true)
    public List<TiendaDto> obtenerTiendasActivas() {
        logger.info("Obteniendo tiendas activas");

        List<Tienda> tiendas = tiendaRepository.findTiendasActivas();
        return tiendas.stream()
                     .map(this::convertirEntidadADto)
                     .collect(Collectors.toList());
    }

    /**
     * Convertir entidad a DTO
     */
    private TiendaDto convertirEntidadADto(Tienda tienda) {
        TiendaDto dto = new TiendaDto();
        dto.setIdTienda(tienda.getIdTienda());
        dto.setCodigoTienda(tienda.getCodigoTienda());
        dto.setNombreTienda(tienda.getNombreTienda());
        dto.setDireccion(tienda.getDireccion());
        dto.setCiudad(tienda.getCiudad());
        dto.setEstado(tienda.getEstado());
        dto.setCodigoPostal(tienda.getCodigoPostal());
        dto.setTelefono(tienda.getTelefono());
        dto.setEmail(tienda.getEmail());
        dto.setGerente(tienda.getGerente());
        dto.setRegion(tienda.getRegion());
        dto.setZona(tienda.getZona());
        dto.setTipoTienda(tienda.getTipoTienda());
        dto.setEstadoTienda(tienda.getEstadoTienda());
        dto.setFechaApertura(tienda.getFechaApertura());
        dto.setFechaCreacion(tienda.getFechaCreacion());
        dto.setFechaModificacion(tienda.getFechaModificacion());
        dto.setUsuarioCreacion(tienda.getUsuarioCreacion());
        dto.setUsuarioModificacion(tienda.getUsuarioModificacion());
        dto.setObservaciones(tienda.getObservaciones());
        return dto;
    }

    /**
     * Convertir DTO a entidad
     */
    private Tienda convertirDtoAEntidad(TiendaDto dto) {
        Tienda tienda = new Tienda();
        tienda.setCodigoTienda(dto.getCodigoTienda());
        tienda.setNombreTienda(dto.getNombreTienda());
        tienda.setDireccion(dto.getDireccion());
        tienda.setCiudad(dto.getCiudad());
        tienda.setEstado(dto.getEstado());
        tienda.setCodigoPostal(dto.getCodigoPostal());
        tienda.setTelefono(dto.getTelefono());
        tienda.setEmail(dto.getEmail());
        tienda.setGerente(dto.getGerente());
        tienda.setRegion(dto.getRegion());
        tienda.setZona(dto.getZona());
        tienda.setTipoTienda(dto.getTipoTienda() != null ? dto.getTipoTienda() : "Sucursal");
        tienda.setEstadoTienda(dto.getEstadoTienda() != null ? dto.getEstadoTienda() : "ACTIVO");
        tienda.setFechaApertura(dto.getFechaApertura());
        tienda.setObservaciones(dto.getObservaciones());
        return tienda;
    }

    /**
     * Actualizar campos de la tienda
     */
    private void actualizarCamposTienda(Tienda tienda, TiendaDto dto) {
        tienda.setCodigoTienda(dto.getCodigoTienda());
        tienda.setNombreTienda(dto.getNombreTienda());
        tienda.setDireccion(dto.getDireccion());
        tienda.setCiudad(dto.getCiudad());
        tienda.setEstado(dto.getEstado());
        tienda.setCodigoPostal(dto.getCodigoPostal());
        tienda.setTelefono(dto.getTelefono());
        tienda.setEmail(dto.getEmail());
        tienda.setGerente(dto.getGerente());
        tienda.setRegion(dto.getRegion());
        tienda.setZona(dto.getZona());
        tienda.setTipoTienda(dto.getTipoTienda());
        tienda.setEstadoTienda(dto.getEstadoTienda());
        tienda.setFechaApertura(dto.getFechaApertura());
        tienda.setObservaciones(dto.getObservaciones());
    }
}

File: facturacion-cibercom/services/tiendaService.ts
// Servicio para la gestión de tiendas
import { apiUrl } from './api';
export interface Tienda {
  idTienda?: number;
  codigoTienda: string;
  nombreTienda: string;
  direccion?: string;
  ciudad?: string;
  estado?: string;
  codigoPostal?: string;
  telefono?: string;
  email?: string;
  gerente?: string;
  region?: string;
  zona?: string;
  tipoTienda?: string;
  estadoTienda?: string;
  fechaApertura?: string;
  fechaCreacion?: string;
  fechaModificacion?: string;
  usuarioCreacion?: string;
  usuarioModificacion?: string;
  observaciones?: string;
}

export interface TiendaResponse {
  success: boolean;
  message: string;
  data?: Tienda | Tienda[];
  total?: number;
  error?: string;
}

export interface EstadisticasTiendas {
  totalTiendas: number;
  tiendasActivas: number;
  tiendasInactivas: number;
  tiendasSuspendidas: number;
  regiones: string[];
  zonas: string[];
  tiposTienda: string[];
}

export interface FiltrosTienda {
  estadoTienda?: string;
  region?: string;
  zona?: string;
  tipoTienda?: string;
  busqueda?: string;
}

class TiendaService {
  private baseUrl = apiUrl('/tiendas');

  /**
   * Crear una nueva tienda
   */
  async crearTienda(tienda: Tienda): Promise<TiendaResponse> {
    try {
      const response = await fetch(this.baseUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(tienda),
      });

      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al crear tienda:', error);
      return {
        success: false,
        message: 'Error de conexión al crear la tienda',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Actualizar una tienda existente
   */
  async actualizarTienda(id: number, tienda: Tienda): Promise<TiendaResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/${id}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(tienda),
      });

      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al actualizar tienda:', error);
      return {
        success: false,
        message: 'Error de conexión al actualizar la tienda',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Listar tiendas con filtros opcionales
   */
  async listarTiendas(filtros?: FiltrosTienda): Promise<TiendaResponse> {
    try {
      const params = new URLSearchParams();
      
      if (filtros) {
        if (filtros.estadoTienda) params.append('estadoTienda', filtros.estadoTienda);
        if (filtros.region) params.append('region', filtros.region);
        if (filtros.zona) params.append('zona', filtros.zona);
        if (filtros.tipoTienda) params.append('tipoTienda', filtros.tipoTienda);
        if (filtros.busqueda) params.append('busqueda', filtros.busqueda);
      }

      const url = params.toString() ? `${this.baseUrl}?${params.toString()}` : this.baseUrl;
      const response = await fetch(url);

      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al listar tiendas:', error);
      return {
        success: false,
        message: 'Error de conexión al obtener las tiendas',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Obtener una tienda por ID
   */
  async obtenerTiendaPorId(id: number): Promise<TiendaResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/${id}`);
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al obtener tienda:', error);
      return {
        success: false,
        message: 'Error de conexión al obtener la tienda',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Obtener una tienda por código
   */
  async obtenerTiendaPorCodigo(codigo: string): Promise<TiendaResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/codigo/${codigo}`);
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al obtener tienda por código:', error);
      return {
        success: false,
        message: 'Error de conexión al obtener la tienda',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Eliminar una tienda (soft delete)
   */
  async eliminarTienda(id: number, usuarioModificacion?: string): Promise<TiendaResponse> {
    try {
      const url = usuarioModificacion ? `${this.baseUrl}/${id}?usuarioModificacion=${encodeURIComponent(usuarioModificacion)}` : `${this.baseUrl}/${id}`;
      const response = await fetch(url, {
        method: 'DELETE',
      });

      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al eliminar tienda:', error);
      return {
        success: false,
        message: 'Error de conexión al eliminar la tienda',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Eliminar una tienda permanentemente
   */
  async eliminarTiendaPermanente(id: number): Promise<TiendaResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/${id}/permanente`, {
        method: 'DELETE',
      });

      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al eliminar tienda permanentemente:', error);
      return {
        success: false,
        message: 'Error de conexión al eliminar la tienda',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Obtener estadísticas de tiendas
   */
  async obtenerEstadisticas(): Promise<{ success: boolean; data?: EstadisticasTiendas; message: string; error?: string }> {
    try {
      const response = await fetch(`${this.baseUrl}/estadisticas`);
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al obtener estadísticas:', error);
      return {
        success: false,
        message: 'Error de conexión al obtener las estadísticas',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Obtener tiendas activas
   */
  async obtenerTiendasActivas(): Promise<TiendaResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/activas`);
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al obtener tiendas activas:', error);
      return {
        success: false,
        message: 'Error de conexión al obtener las tiendas activas',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<TiendaResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/health`);
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error al verificar salud del servicio de tiendas:', error);
      return {
        success: false,
        message: 'Error de conexión al verificar salud del servicio de tiendas',
        error: error instanceof Error ? error.message : 'Error desconocido'
      };
    }
  }
}

export const tiendaService = new TiendaService();